File: Core/src/main/java/com/hqq/core/ui/vm/BaseListActivity.java
package com.hqq.core.ui.vm;

import android.view.View;
import android.view.ViewGroup;

import androidx.databinding.ViewDataBinding;
import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.hqq.core.R;
import com.hqq.core.ui.model.BaseListModel;

import java.util.ArrayList;

/**
 * @Author : huangqiqiang
 * @Package : com.hqq.core.ui.vm
 * @FileName :   BaseListActivity
 * @Date : 2020/7/29 0029  上午 9:54
 * @Email : qiqiang213@gmail.com
 * @Descrive :
 */
public abstract class BaseListActivity<T extends ViewDataBinding, K extends BaseListViewModel, AD extends BaseQuickAdapter>
        extends BaseActivity<T, K> implements BaseListModel.IBaseListModelView<AD> {

    protected RecyclerView mRcList;
    protected AD mAdapter;
    protected RecyclerView.LayoutManager mLayoutManager;
    protected BaseListModel mBaseListModel;


    @Override
    public int getLayoutId() {
        return R.layout.activity_recycle_view;
    }

    @Override
    public void initViews() {
        mBaseListModel = new BaseListModel(this, this);
        mLayoutManager = getRcLayoutManager();
        mAdapter = initAdapter();

        mRcList = mBaseListModel.checkRecycleView(mRcList, mRootViewBuild.getRootView());

        mBaseListModel.initRecycleView(mRcList, mAdapter, mLayoutManager);

        mBaseListModel.initPtrPullDown(mRootViewBuild.getRootView());

        mViewModel.mDate.observe(this, new Observer<ArrayList>() {
            @Override
            public void onChanged(ArrayList arrayList) {
                mBaseListModel.fillingData(arrayList);

            }
        });


        initData();

    }

    @Override
    public void onLoadMoreRequested() {
        mViewModel.onLoadMore();
    }

    @Override
    public boolean isShowLoadMore() {
        return false;
    }

    @Override
    public AD getAdapter() {
        return mAdapter;
    }

    @Override
    public int getPageCount() {
        return mViewModel.getPageCount();
    }

    @Override
    public int getPageSize() {
        return mViewModel.getPageSize();
    }

    @Override
    public void addPageCount() {
        mViewModel.setPageCount(mViewModel.getPageCount() + 1);
    }

    @Override
    public void onRefreshBegin() {
        mViewModel.setPageCount(1);
        mAdapter.loadMoreComplete();
        mViewModel.onLoadMore();
    }

    @Override
    public ViewGroup getListView() {
        return mRcList;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mBaseListModel.onDestroy();
        mBaseListModel = null;
    }

    @Override
    public RecyclerView.LayoutManager getRcLayoutManager() {
        return new LinearLayoutManager(this);
    }


    @Override
    public void onItemClick(BaseQuickAdapter adapter, View view, int position) {

    }

    @Override
    public void onItemChildClick(BaseQuickAdapter adapter, View view, int position) {
    }


}

File: Core/src/main/java/com/easy/core/ui/list3/BaseListActivity.kt
package com.easy.core.ui.list3

import android.view.View
import android.view.ViewGroup
import androidx.annotation.CallSuper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.chad.library.adapter.base.module.LoadMoreModule
import com.easy.core.BaseCommonsKey
import com.easy.core.ui.base.BaseActivity

/**
 * @Author : huangqiqiang
 * @Package : com.core.library
 * @FileName :   BaseListActivity
 * @Date : 2018/11/5 0005  下午 4:23
 * @Describe :
 * @Email :  qiqiang213@gmail.com
 */
abstract class BaseListActivity : BaseActivity(), BaseListModel3.IBaseListModelView3 {

    /**
     *  分页大小 默认是20 可以在Config 中统一配置
     */
    override var pageSize = BaseCommonsKey.PAGE_SIZE

    /**
     *  分页下标
     */
    override var pageCount = 1

    /**
     *  正常List 默认使用 自带的ListView  如果页面包含其他内容 则使用 LayoutViewId
     */
    override fun getLayoutViewId(): Int {
        return 0
    }
    /**
     *  RecycleView Manager
     */
    override var layoutManager: RecyclerView.LayoutManager = LinearLayoutManager(this)

    /**
     *  RecycleView 对象
     */
    override var listView: RecyclerView? = null

    /**
     * 重写了 setViewId    执行 setRootView
     * 默认都会执行一般这个方法 但是没有添加到rooView中
     *
     * @return
     */
    override fun getLayoutView(group: ViewGroup): View? {
        return if (getLayoutViewId() <= 0) {
            BaseListModel3.createRecycleView(this)
        } else {
            null
        }
    }

    override lateinit var listModel: BaseListModel3

    /**
     * 初始化  基础数据
     */
    override fun initView() {
        listModel = BaseListModel3(this, rootViewImpl)
        initData()
    }

    /**
     *  分页增加
     */
    override fun addPageCount() {
        pageCount += 1
    }

    /**
     *  下拉刷新 开始从第一页 获取数据
     */
    override fun onRefreshBegin() {
        if (adapter is LoadMoreModule) {

            pageCount = 1
            adapter.loadMoreModule.loadMoreComplete()
            onLoadMore()
        }
    }

    /**
     *  销毁
     */
    @CallSuper
    override fun onDestroy() {
        super.onDestroy()
    }

    /**
     *  加载更多数据
     */
    override fun onLoadMore() {
    }


}